handlers/stream_cache.go

```diff
@@ -0,0 +1,117 @@
+package handlers
+
+import (
+	"fmt"
+	"sync"
+)
+
+// Shared server state. Each pipeline (usually endpoint) having separate books.
+// State is protected by mutex allowing concurent usage.
+// All state manipulation is contained in this file with goal to be brief and release mutex asap.
+type StreamCache struct {
+	Segmenting  SegmentingCache
+	Transcoding TranscodingCache
+}
+
+// Returns pointer so each handler would refer to same object (kind of singleton)
+func NewStreamCache() *StreamCache {
+	c := &StreamCache{}
+	c.Init()
+	return c
+}
+
+type TranscodingCache struct {
+	pushes map[string]*SegmentInfo
+	mutex  sync.Mutex
+}
+
+type SegmentInfo struct {
+	CallbackUrl   string
+	Source        string   // S3 input we are transcoding
+	UploadDir     string   // S3 destination url for multiple renditions
+	Destionations []string // Rendition URLS go here on push start and removed on push end
+}
+
+func (c *TranscodingCache) Init() {
+	c.pushes = make(map[string]*SegmentInfo)
+}
+
+type Empty = bool
+
+func (c *TranscodingCache) RemovePushDestination(streamName, destination string) Empty {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	info, ok := c.pushes[streamName]
+	if ok {
+		for i := 0; i < len(info.Destionations); i++ {
+			if info.Destionations[i] == destination {
+				info.Destionations[i] = info.Destionations[len(info.Destionations)-1]
+				info.Destionations = info.Destionations[:len(info.Destionations)-1]
+				break
+			}
+		}
+		return len(info.Destionations) == 0
+	}
+	return false
+}
+
+func (c *TranscodingCache) Get(streamName string) (SegmentInfo, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	info, ok := c.pushes[streamName]
+	if ok {
+		return *info, nil
+	}
+	return SegmentInfo{}, fmt.Errorf("cache mismatch for %s", streamName)
+}
+
+func (c *TranscodingCache) AddDestination(streamName, destination string) {
+	c.mutex.Lock()
+	info, ok := c.pushes[streamName]
+	if ok {
+		info.Destionations = append(info.Destionations, destination)
+	}
+	c.mutex.Unlock()
```

Comment:
Should this lock be deferred?

Reply:
hm, is there any difference? i am not familiar with panic yet